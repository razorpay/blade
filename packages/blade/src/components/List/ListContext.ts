import React from 'react';
import type { ListProps } from './List';
import type { IconComponent } from '~components/Icons';

export type ListContextType = {
  level?: number;
  size: NonNullable<ListProps['size']>;
  icon?: IconComponent;
  variant: NonNullable<ListProps['variant']>;
};

const ListContext = React.createContext<ListContextType>({
  level: undefined,
  size: 'medium',
  variant: 'unordered',
});
const ListProvider = ListContext.Provider;

const useListContext = (): ListContextType => {
  const context = React.useContext(ListContext);
  return context;
};

export { useListContext, ListProvider };
