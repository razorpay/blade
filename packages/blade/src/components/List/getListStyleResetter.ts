import type { CSSObject } from 'styled-components';

const getListStyleResetter = (): CSSObject => {
  return {
    listStyleType: 'none',
    padding: '0px',
    margin: '0px',
  };
};

export { getListStyleResetter };
