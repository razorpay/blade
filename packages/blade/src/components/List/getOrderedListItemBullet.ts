/**
 * Converts number to roman number
 *
 */
const romanize = (number: number): string => {
  const romanKeys: Record<string, number> = {
    m: 1000,
    cm: 900,
    d: 500,
    cd: 400,
    c: 100,
    xc: 90,
    l: 50,
    xl: 40,
    x: 10,
    ix: 9,
    v: 5,
    iv: 4,
    i: 1,
  };
  let result = '';

  Object.keys(romanKeys).forEach((key) => {
    result += key.repeat(Math.floor(number / romanKeys[key]));
    number %= romanKeys[key];
  });

  return result;
};

/**
 * Converts number to alphabet. After 26 letters, the output will contain 2 alphabets eg) aa, ab, ac,...
 *
 */
const alphabetize = (number: number): string => {
  // Decrement the input number since array starts with 0. Eg) 'a' will be at 0th position and not 1st position in the array.
  const indexedNumber = --number;
  const alphabets = 'abcdefghijklmnopqrstuvwxyz';
  const firstAlphabets = indexedNumber >= 26 ? alphabetize(Math.floor(indexedNumber / 26) - 1) : '';
  const lastAlphabet = alphabets[indexedNumber % 26];
  return `${firstAlphabets}${lastAlphabet}`;
};

/**
 * Get the list item bullet number/roman number/alphabet for the passed level and list item number
 *
 */
const getOrderedListItemBullet = ({
  itemNumber,
  level,
}: {
  itemNumber: number;
  level: number;
}): string => {
  if (isNaN(itemNumber)) return '';
  switch (level) {
    case 1:
      return `${itemNumber}`;
    case 2:
      return `${alphabetize(itemNumber)}`;
    case 3:
      return `${romanize(itemNumber)}`;
    default:
      return `${itemNumber}`;
  }
};

export { getOrderedListItemBullet };
