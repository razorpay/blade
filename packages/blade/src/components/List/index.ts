export { List } from './List';
export type { ListProps } from './List';
export { ListItem } from './ListItem';
export type { ListItemProps } from './ListItem';
