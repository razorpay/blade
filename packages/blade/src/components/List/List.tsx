import React, { useMemo } from 'react';
import styled from 'styled-components';
import { ListProvider, useListContext } from './ListContext';
import { UnorderedList } from './UnorderedList';
import { OrderedList } from './OrderedList';
import type { ListItemProps } from './ListItem';
import getIn from '~utils/lodashButBetter/get';
import type { IconComponent } from '~components/Icons';
import { metaAttribute, MetaConstants } from '~utils/metaAttribute';
import type { DotNotationSpacingStringToken, TestID } from '~utils/types';
import BaseBox from '~components/Box/BaseBox';
import { getStyledProps } from '~components/Box/styledProps';
import type { StyledPropsBlade } from '~components/Box/styledProps';
import { assignWithoutSideEffects } from '~utils/assignWithoutSideEffects';
import { makeSpace } from '~utils/makeSpace';
import { makeAccessible } from '~utils/makeAccessible';
import { isValidAllowedChildren } from '~utils/isValidAllowedChildren';
import { throwBladeError } from '~utils/logger';

type ListCommonProps = {
  /**
   * ListItem to be rendered for the List.
   *
   */
  children: React.ReactElement<ListItemProps> | React.ReactElement<ListItemProps>[];
  /**
   * Sets the variant to be rendered for the List.
   *
   * @default 'unordered'
   */
  variant?: 'unordered' | 'ordered' | 'ordered-filled';
  /**
   * Sets the size for the List.
   *
   * @default 'medium'
   */
  size?: 'small' | 'medium' | 'large';
} & TestID &
  StyledPropsBlade;

type ListWithIconProps = ListCommonProps & {
  variant?: 'unordered';
  icon?: IconComponent;
};
type ListWithoutIconProps = ListCommonProps & {
  variant?: 'ordered' | 'ordered-filled';
  icon?: undefined;
};
type ListProps = ListWithIconProps | ListWithoutIconProps;

const StyledOrderedList = styled(OrderedList)<{ marginTop?: DotNotationSpacingStringToken }>(
  ({ marginTop, theme }) => ({
    marginTop: marginTop ? makeSpace(getIn(theme, marginTop)) : undefined,
  }),
);

const StyledUnorderedList = styled(UnorderedList)<{ marginTop?: DotNotationSpacingStringToken }>(
  ({ marginTop, theme }) => ({
    marginTop: marginTop ? makeSpace(getIn(theme, marginTop)) : undefined,
  }),
);

/**
 * List Component is used to display a set of related items that are composed of text/links. Each list item begins with a bullet or a number.
 *
 * ## Usage
 *
 * ```tsx
 *   <List
 *     variant='unordered'
 *     size='medium'
 *   >
 *    <ListItem>
 *      Item 1
 *      <List>
 *        <ListItem>Item 1.1</ListItem>
 *      </List>
 *    </ListItem>
 *    <ListItem>Item 2</ListItem>
 *  <List />
 * ```
 */
const _List = ({
  variant = 'unordered',
  size,
  children,
  icon,
  testID,
  ...styledProps
}: ListProps): React.ReactElement => {
  const ListElement = variant === 'unordered' ? StyledUnorderedList : StyledOrderedList;
  const { level, size: listContextSize } = useListContext();
  const listContextValue = useMemo(
    () => ({
      level: level ? level + 1 : 1,
      size: size ?? listContextSize,
      icon,
      variant,
    }),
    [icon, level, listContextSize, size, variant],
  );

  const childrenArray = React.Children.toArray(children);
  const childListItems = childrenArray.filter((child) => {
    if (isValidAllowedChildren(child, MetaConstants.ListItem)) {
      return child;
    }
    if (__DEV__) {
      throwBladeError({
        message: 'You can only pass a ListItem as a child to List.',
        moduleName: 'List',
      });
    }
    return null;
  });

  return (
    <ListProvider value={listContextValue}>
      <BaseBox {...getStyledProps(styledProps)}>
        <ListElement
          {...metaAttribute({ name: MetaConstants.List, testID })}
          {...makeAccessible({ role: 'list' })} // Role needed for react-native
        >
          {variant === 'unordered'
            ? childListItems
            : childListItems.map(
                (child, index) =>
                  React.cloneElement(child as React.ReactElement, { _itemNumber: index + 1 }), // adds _itemNumber for rendering ordered list bullets
              )}
        </ListElement>
      </BaseBox>
    </ListProvider>
  );
};

const List = assignWithoutSideEffects(_List, { componentId: MetaConstants.List });

export { List };
export type { ListProps };
