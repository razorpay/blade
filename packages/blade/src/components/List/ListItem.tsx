import styled from 'styled-components';
import getIn from 'lodash/get';
import React from 'react';
import { Text } from '../Typography';
import type { IconComponent } from '../Icons';
import { useTheme } from '../BladeProvider';
import { useListContext } from './ListContext';
import { UnorderedItemIcon } from './ListItemIcons';
import { ListItemElement } from './ListItemElement';
import {
  listItemBulletMarginRight,
  listItemBulletMarginTop,
  listItemOrderedBulletBoxSize,
  listItemMarginBottom,
  listItemMarginLeft,
} from './listTokens';
import type { ListProps } from './List';
import { getOrderedListItemBullet } from './getOrderedListItemBullet';
import BaseBox from '~components/Box/BaseBox';
import { metaAttribute, MetaConstants } from '~utils/metaAttribute';
import type { TestID } from '~utils/types';
import { assignWithoutSideEffects } from '~utils/assignWithoutSideEffects';
import { getComponentId, isValidAllowedChildren } from '~utils/isValidAllowedChildren';
import { getPlatformType } from '~utils/getPlatformType';
import { throwBladeError } from '~utils/logger';

type ListItemProps = {
  /**
   * Children to be rendered for ListItem. This can be a text, ListItemLink or another List.
   *
   */
  children: React.ReactNode;
  /**
   * Icon to be rendered for a ListItem's bullet.
   *
   */
  icon?: IconComponent;
  /**
   * This is a private prop to be used only for internal logic purposes.
   *
   */
  _itemNumber?: undefined;
} & TestID;

const StyledListItem = styled(ListItemElement)<{
  level?: number;
  variant: NonNullable<ListProps['variant']>;
  hasIcon: boolean;
}>(({ level, theme, variant, hasIcon }) => ({
  marginLeft: level
    ? getIn(
        theme,
        listItemMarginLeft[
          `${variant}${variant === 'unordered' && hasIcon ? 'WithIcon' : ''}` as NonNullable<
            ListProps['variant'] | 'unorderedWithIcon'
          >
        ][level],
      )
    : 0,
}));

const ListItemContentChildren = ({
  children,
  size,
}: {
  children: React.ReactNode[];
  size: NonNullable<ListProps['size']>;
}): React.ReactElement => {
  /* Having a <View><Text>...</Text><View/> inside <Text /> breaks vertical alignment. Issue: https://github.com/facebook/react-native/issues/31955
    As a workaround, we wrap individual strings in their own <Text /> and handle alignment with a parent <View> (BaseBox).
   */
  return getPlatformType() === 'react-native' ? (
    <BaseBox display="flex" flexDirection="row" flexWrap="wrap">
      {children.map((child, index) => {
        if (typeof child === 'string') {
          return (
            <Text key={index} variant="body" size={size}>
              {child}
            </Text>
          );
        }
        return child;
      })}
    </BaseBox>
  ) : (
    <Text variant="body" size={size}>
      {children}
    </Text>
  );
};

const _ListItem = ({
  children,
  icon: Icon,
  _itemNumber,
  testID,
}: ListItemProps): React.ReactElement => {
  const { level, size, icon: ListContextIcon, variant } = useListContext();
  const { theme, platform } = useTheme();
  const ItemIcon = Icon ?? ListContextIcon;

  if (__DEV__) {
    if (level && level > 3) {
      throwBladeError({
        message: 'List Nesting is allowed only upto 3 levels.',
        moduleName: 'List',
      });
    }
  }

  const childrenArray = React.Children.toArray(children);

  // Get children that are not a List component and are valid allowed children
  const validChildItem = childrenArray.filter((child) => {
    if (getComponentId(child) === MetaConstants.List) return null;

    if (
      typeof child === 'string' ||
      isValidAllowedChildren(child, MetaConstants.ListItemLink) ||
      isValidAllowedChildren(child, MetaConstants.ListItemText) ||
      isValidAllowedChildren(child, MetaConstants.ListItemCode)
    ) {
      return child;
    } else if (__DEV__) {
      throwBladeError({
        message: `You can only pass a List, ListItemLink, ListItemCode, ListItemText or a string as a child to ListItem.`,
        moduleName: 'ListItem',
      });
    }
    return null;
  });
  // Get child that is a List component
  const childList = childrenArray.filter((child) =>
    getComponentId(child) === MetaConstants.List ? child : null,
  );
  const hasIcon = Boolean(ItemIcon);

  return (
    <StyledListItem
      level={level}
      variant={variant}
      hasIcon={hasIcon}
      {...metaAttribute({ name: MetaConstants.ListItem, testID })}
    >
      <BaseBox
        display="flex"
        flexDirection="row"
        alignItems="center"
        marginBottom={listItemMarginBottom}
      >
        {variant === 'unordered' ? (
          <BaseBox
            marginRight={listItemBulletMarginRight[variant]}
            marginTop={
              listItemBulletMarginTop[`${variant}${hasIcon ? 'WithIcon' : ''}`][platform][size]
            }
            display="flex"
            alignSelf="flex-start"
          >
            {ItemIcon ? (
              <ItemIcon size={size} color="surface.text.subdued.lowContrast" />
            ) : (
              <UnorderedItemIcon level={level} />
            )}
          </BaseBox>
        ) : (
          <BaseBox
            width={listItemOrderedBulletBoxSize[variant][platform][size]}
            height={listItemOrderedBulletBoxSize[variant][platform][size]}
            marginRight={listItemBulletMarginRight[variant]}
            marginTop={listItemBulletMarginTop[variant][platform][size]}
            display="flex"
            flexShrink={0}
            justifyContent="center"
            alignSelf="flex-start"
            alignItems="center"
            borderRadius={variant === 'ordered-filled' ? 'max' : undefined}
            backgroundColor={
              variant === 'ordered-filled'
                ? getIn(theme, 'colors.brand.gray.a100.lowContrast')
                : undefined
            }
          >
            <Text variant="body" type="subtle" size={variant === 'ordered' ? size : 'xsmall'}>
              {`${getOrderedListItemBullet({
                itemNumber: _itemNumber ?? 1,
                level: level ?? 1,
              })}${variant === 'ordered' ? '.' : ''}`}
            </Text>
          </BaseBox>
        )}
        <ListItemContentChildren size={size}>{validChildItem}</ListItemContentChildren>
      </BaseBox>
      {childList}
    </StyledListItem>
  );
};

const ListItem = assignWithoutSideEffects(_ListItem, { componentId: MetaConstants.ListItem });

export { ListItem };
export type { ListItemProps };
