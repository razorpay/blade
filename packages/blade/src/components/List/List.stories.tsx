import type { ComponentStory, Meta } from '@storybook/react';
import { Title } from '@storybook/addon-docs';
import type { ReactElement } from 'react';
import { BookmarkIcon } from '../Icons';
import { Heading } from '../Typography';
import type { ListProps } from './List';
import { List } from './List';
import { ListItem } from './ListItem';
import { ListItemLink } from './ListItemLink';
import { ListItemCode } from './ListItemCode';
import iconMap from '~components/Icons/iconMap';
import { Sandbox } from '~src/_helpers/storybook/Sandbox';
import StoryPageWrapper from '~src/_helpers/storybook/StoryPageWrapper';
import BaseBox from '~components/Box/BaseBox';

const Page = (): ReactElement => {
  return (
    <StoryPageWrapper
      componentDescription="List displays a set of related items that are composed of text/links."
      componentName="List"
      figmaURL={{
        paymentTheme:
          'https://www.figma.com/file/jubmQL9Z8V7881ayUD95ps/Blade---Payment-Light?node-id=23205%3A446859&t=itEw2V8u5Q0PPGJq-4',
        bankingTheme:
          'https://www.figma.com/file/sAdplk2uYnI2ILnDKUxycW/Blade---Banking-Dark?node-id=13864%3A436458&t=nNhw3mY86j85bYfl-4',
      }}
    >
      <Title>Usage</Title>
      <Sandbox>
        {`
          import { List, ListItem } from '@razorpay/blade/components';

          function App(): JSX.Element {
            return (
              <List>
                <ListItem>
                  Level 1
                  <List>
                    <ListItem>
                      Level 2
                      <List>
                        <ListItem>
                          Level 3
                        </ListItem>
                      </List>
                    </ListItem>
                  </List>
                </ListItem>
              </List>
            )
          }

          export default App;
        `}
      </Sandbox>
    </StoryPageWrapper>
  );
};

export default {
  title: 'Components/List',
  component: List,
  args: {
    size: 'medium',
    variant: 'unordered',
  },
  argTypes: {
    icon: {
      name: 'icon',
      type: 'select',
      options: Object.keys(iconMap),
      mapping: iconMap,
    },
  },
  parameters: {
    docs: {
      page: Page,
    },
  },
} as Meta<ListProps>;

const ListTemplate: ComponentStory<typeof List> = ({ ...args }) => {
  return (
    <List variant="unordered" {...args}>
      <ListItem>
        Debit Card
        <List variant="unordered" {...args}>
          <ListItem>
            HDFC
            <List variant="unordered" {...args}>
              <ListItem>Domestic</ListItem>
              <ListItem>International</ListItem>
            </List>
          </ListItem>
        </List>
      </ListItem>
      <ListItem>
        Credit Card
        <List variant="unordered" {...args}>
          <ListItem>
            ICICI
            <List variant="unordered" {...args}>
              <ListItem>Domestic</ListItem>
              <ListItem>International</ListItem>
            </List>
          </ListItem>
        </List>
      </ListItem>
      <ListItem>
        Netbanking
        <List variant="unordered" {...args}>
          <ListItem>HDFC</ListItem>
          <ListItem>ICICI</ListItem>
        </List>
      </ListItem>
    </List>
  );
};

export const Default = ListTemplate.bind({});
// Need to do this because of storybook's weird naming convention, More details here: https://storybook.js.org/docs/react/writing-stories/naming-components-and-hierarchy#single-story-hoisting
Default.storyName = 'Default';
Default.args = {};

const ListMixNestedTemplate: ComponentStory<typeof List> = () => {
  return (
    <List variant="ordered">
      <ListItem>
        Debit Card
        <List variant="unordered">
          <ListItem>
            HDFC
            <List variant="ordered">
              <ListItem>Domestic</ListItem>
              <ListItem>International</ListItem>
            </List>
          </ListItem>
        </List>
      </ListItem>
      <ListItem>
        Credit Card
        <List variant="unordered">
          <ListItem>
            ICICI
            <List variant="ordered">
              <ListItem>Domestic</ListItem>
              <ListItem>International</ListItem>
            </List>
          </ListItem>
        </List>
      </ListItem>
      <ListItem>
        Netbanking
        <List variant="unordered">
          <ListItem>HDFC</ListItem>
          <ListItem>ICICI</ListItem>
        </List>
      </ListItem>
    </List>
  );
};

export const ListMixNested = ListMixNestedTemplate.bind({});
// Need to do this because of storybook's weird naming convention, More details here: https://storybook.js.org/docs/react/writing-stories/naming-components-and-hierarchy#single-story-hoisting
ListMixNested.storyName = 'Unordered & Ordered Mix';

const ListWithSizesTemplate: ComponentStory<typeof List> = ({ ...args }) => {
  return (
    <BaseBox>
      <Heading>Small Size:</Heading>
      <List {...args} size="small">
        <ListItem>
          Level 1
          <List {...args} size="small">
            <ListItem>
              Level 2
              <List {...args} size="small">
                <ListItem>Level 3</ListItem>
              </List>
            </ListItem>
          </List>
        </ListItem>
      </List>
      <Heading>Medium Size:</Heading>
      <List {...args} size="medium">
        <ListItem>
          Level 1
          <List {...args} size="medium">
            <ListItem>
              Level 2
              <List {...args} size="medium">
                <ListItem>Level 3</ListItem>
              </List>
            </ListItem>
          </List>
        </ListItem>
      </List>
    </BaseBox>
  );
};

export const UnorderedListWithSizes = ListWithSizesTemplate.bind({});
UnorderedListWithSizes.storyName = 'Unordered - Sizes';
UnorderedListWithSizes.parameters = {
  controls: {
    disable: true,
  },
};
UnorderedListWithSizes.args = {
  variant: 'unordered',
};

export const OrderedListWithSizes = ListWithSizesTemplate.bind({});
OrderedListWithSizes.storyName = 'Ordered - Sizes';
OrderedListWithSizes.parameters = {
  controls: {
    disable: true,
  },
};
OrderedListWithSizes.args = {
  variant: 'ordered',
};

const OrderedFilledListWithSizesTemplate: ComponentStory<typeof List> = () => {
  return (
    <BaseBox>
      <Heading>Small Size:</Heading>
      <List variant="ordered-filled" size="small">
        <ListItem>
          <ListItemLink>Build Integration:</ListItemLink> Use the sample codes to integrate the
          Razorpay Web Standard Checkout on your website.
        </ListItem>
        <ListItem>
          <ListItemLink>Test Integration:</ListItemLink> Test the integration to ensure it was
          successful.
        </ListItem>
        <ListItem>
          <ListItemLink>Go-live Checklist:</ListItemLink> Check the go-live checklist before taking
          the integration live.
        </ListItem>
      </List>
      <Heading>Medium Size:</Heading>
      <List variant="ordered-filled" size="medium">
        <ListItem>
          <ListItemLink>Build Integration:</ListItemLink> Use the sample codes to integrate the
          Razorpay Web Standard Checkout on your website.
        </ListItem>
        <ListItem>
          <ListItemLink>Test Integration:</ListItemLink> Test the integration to ensure it was
          successful.
        </ListItem>
        <ListItem>
          <ListItemLink>Go-live Checklist:</ListItemLink> Check the go-live checklist before taking
          the integration live.
        </ListItem>
      </List>
    </BaseBox>
  );
};

export const OrderedFilledListWithSizes = OrderedFilledListWithSizesTemplate.bind({});
OrderedFilledListWithSizes.storyName = 'OrderedFilled - Sizes';

const ListWithLinkAndIconTemplate: ComponentStory<typeof List> = () => {
  return (
    <List variant="unordered" icon={BookmarkIcon}>
      <ListItem>
        <ListItemLink>Troubleshooting and FAQs</ListItemLink>
      </ListItem>
      <ListItem>
        <ListItemLink>Payment Methods</ListItemLink>
      </ListItem>
      <ListItem>
        <ListItemLink>International Currency Support</ListItemLink>
      </ListItem>
      <ListItem>
        <ListItemLink>Bank Downtime</ListItemLink>
      </ListItem>
    </List>
  );
};

export const ListWithLinkAndIcon = ListWithLinkAndIconTemplate.bind({});
ListWithLinkAndIcon.storyName = 'Link & Icon';

const ListWithCodeTemplate: ComponentStory<typeof List> = () => {
  return (
    <BaseBox>
      <Heading>Small Size:</Heading>
      <List variant="ordered" size="small">
        <ListItem>
          Bump blade version to <ListItemCode>v6.0.0</ListItemCode>
        </ListItem>
        <ListItem>
          Run <ListItemCode>yarn install</ListItemCode>
        </ListItem>

        <ListItem>
          Run <ListItemCode>yarn start</ListItemCode>
        </ListItem>
      </List>
      <Heading>Medium Size:</Heading>
      <List variant="ordered" size="medium">
        <ListItem>
          Bump blade version to <ListItemCode>v6.0.0</ListItemCode>
        </ListItem>
        <ListItem>
          Run <ListItemCode>yarn install</ListItemCode>
        </ListItem>

        <ListItem>
          Run <ListItemCode>yarn start</ListItemCode>
        </ListItem>
      </List>
    </BaseBox>
  );
};
export const ListWithCodeAndIcon = ListWithCodeTemplate.bind({});
ListWithCodeAndIcon.storyName = 'With Inline Code';
