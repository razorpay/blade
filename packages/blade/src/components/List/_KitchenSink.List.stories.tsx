import { composeStories } from '@storybook/react';
import * as listStories from './List.stories';
import { Box } from '~components/Box';
import { Heading } from '~components/Typography';

const allStories = Object.values(composeStories(listStories));

export const List = (): JSX.Element => {
  return (
    <Box display="flex" flexDirection="column" gap="spacing.4">
      {allStories.map((Story) => {
        return (
          <>
            <Heading>{Story.storyName}</Heading>
            <Story />
          </>
        );
      })}
    </Box>
  );
};

export default {
  title: 'Components/KitchenSink/List',
  component: List,
  parameters: {
    // enable Chromatic's snapshotting only for kitchensink
    chromatic: { disableSnapshot: false },
    options: { showPanel: false },
  },
};
