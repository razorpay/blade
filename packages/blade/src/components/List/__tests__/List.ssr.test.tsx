import { List } from '../List';
import { ListItem } from '../ListItem';
import renderWithSSR from '~utils/testing/renderWithSSR.web';

describe('<List />', () => {
  it('should render List with default properties', () => {
    const { container, queryAllByRole, getByText } = renderWithSSR(
      <List>
        <ListItem>
          Level 1
          <List>
            <ListItem>
              Level 2
              <List>
                <ListItem>Level 3</ListItem>
              </List>
            </ListItem>
          </List>
        </ListItem>
      </List>,
    );
    expect(container).toMatchSnapshot();
    expect(queryAllByRole('list')).toHaveLength(3);
    expect(queryAllByRole('listitem')).toHaveLength(3);
    expect(getByText('Level 1')).toBeInTheDocument();
    expect(getByText('Level 2')).toBeInTheDocument();
    expect(getByText('Level 3')).toBeInTheDocument();
  });
});
