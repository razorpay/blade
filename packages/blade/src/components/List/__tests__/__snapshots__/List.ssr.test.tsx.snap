// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> should render List with default properties 1`] = `"<div id=\\"root\\" data-reactroot=\\"\\"><ul data-blade-component=\\"list\\" role=\\"list\\" class=\\"UnorderedListweb__UnorderedList-sc-1tb6ajo-0 List__StyledUnorderedList-bqf6r7-1 dbSpGu erLmOF\\"><li data-blade-component=\\"list-item\\" class=\\"ListItemElementweb__ListItemElement-sc-2pr284-0 ListItem__StyledListItem-sc-6bxu4b-0  bALAZM\\"><div display=\\"flex\\" class=\\"Boxweb__Box-xvyj85-0 jlsgap\\"><div display=\\"flex\\" class=\\"Boxweb__Box-xvyj85-0 cXVkPc\\"><svg aria-hidden=\\"true\\" data-blade-component=\\"icon\\" height=\\"6px\\" viewBox=\\"0 0 6 6\\" width=\\"6px\\" fill=\\"none\\"><circle cx=\\"3px\\" cy=\\"3px\\" r=\\"3px\\" fill=\\"hsla(214, 18%, 69%, 1)\\"></circle></svg></div><p color=\\"surface.text.normal.lowContrast\\" font-family=\\"text\\" font-size=\\"100\\" font-weight=\\"regular\\" font-style=\\"normal\\" class=\\"BaseTextweb__StyledBaseText-sc-1ob51ew-0 dxOnav Text__StyledText-wkh56f-0\\" data-blade-component=\\"text\\">Level 1</p></div><ul data-blade-component=\\"list\\" role=\\"list\\" class=\\"UnorderedListweb__UnorderedList-sc-1tb6ajo-0 List__StyledUnorderedList-bqf6r7-1 dbSpGu\\"><li data-blade-component=\\"list-item\\" class=\\"ListItemElementweb__ListItemElement-sc-2pr284-0 ListItem__StyledListItem-sc-6bxu4b-0  hWTYHD\\"><div display=\\"flex\\" class=\\"Boxweb__Box-xvyj85-0 jlsgap\\"><div display=\\"flex\\" class=\\"Boxweb__Box-xvyj85-0 cXVkPc\\"><svg aria-hidden=\\"true\\" data-blade-component=\\"icon\\" height=\\"6px\\" viewBox=\\"0 0 6 6\\" width=\\"6px\\" fill=\\"none\\"><circle cx=\\"3px\\" cy=\\"3px\\" r=\\"2.5px\\" stroke=\\"hsla(214, 18%, 69%, 1)\\"></circle></svg></div><p color=\\"surface.text.normal.lowContrast\\" font-family=\\"text\\" font-size=\\"100\\" font-weight=\\"regular\\" font-style=\\"normal\\" class=\\"BaseTextweb__StyledBaseText-sc-1ob51ew-0 dxOnav Text__StyledText-wkh56f-0\\" data-blade-component=\\"text\\">Level 2</p></div><ul data-blade-component=\\"list\\" role=\\"list\\" class=\\"UnorderedListweb__UnorderedList-sc-1tb6ajo-0 List__StyledUnorderedList-bqf6r7-1 dbSpGu\\"><li data-blade-component=\\"list-item\\" class=\\"ListItemElementweb__ListItemElement-sc-2pr284-0 ListItem__StyledListItem-sc-6bxu4b-0  hWTYHD\\"><div display=\\"flex\\" class=\\"Boxweb__Box-xvyj85-0 jlsgap\\"><div display=\\"flex\\" class=\\"Boxweb__Box-xvyj85-0 cXVkPc\\"><svg aria-hidden=\\"true\\" data-blade-component=\\"icon\\" height=\\"6px\\" viewBox=\\"0 0 6 6\\" width=\\"6px\\" fill=\\"none\\"><rect height=\\"6px\\" width=\\"6px\\" rx=\\"1px\\" fill=\\"hsla(214, 18%, 69%, 1)\\"></rect></svg></div><p color=\\"surface.text.normal.lowContrast\\" font-family=\\"text\\" font-size=\\"100\\" font-weight=\\"regular\\" font-style=\\"normal\\" class=\\"BaseTextweb__StyledBaseText-sc-1ob51ew-0 dxOnav Text__StyledText-wkh56f-0\\" data-blade-component=\\"text\\">Level 3</p></div></li></ul></li></ul></li></ul></div>"`;

exports[`<List /> should render List with default properties 2`] = `
<div
  data-reactroot=""
  id="root"
>
  <ul
    class="UnorderedListweb__UnorderedList-sc-1tb6ajo-0 List__StyledUnorderedList-bqf6r7-1 dbSpGu erLmOF"
    data-blade-component="list"
    role="list"
  >
    <li
      class="ListItemElementweb__ListItemElement-sc-2pr284-0 ListItem__StyledListItem-sc-6bxu4b-0  bALAZM"
      data-blade-component="list-item"
    >
      <div
        class="Boxweb__Box-xvyj85-0 jlsgap"
        display="flex"
      >
        <div
          class="Boxweb__Box-xvyj85-0 cXVkPc"
          display="flex"
        >
          <svg
            aria-hidden="true"
            data-blade-component="icon"
            fill="none"
            height="6px"
            viewBox="0 0 6 6"
            width="6px"
          >
            <circle
              cx="3px"
              cy="3px"
              fill="hsla(214, 18%, 69%, 1)"
              r="3px"
            />
          </svg>
        </div>
        <p
          class="BaseTextweb__StyledBaseText-sc-1ob51ew-0 dxOnav Text__StyledText-wkh56f-0"
          color="surface.text.normal.lowContrast"
          data-blade-component="text"
          font-family="text"
          font-size="100"
          font-style="normal"
          font-weight="regular"
        >
          Level 1
        </p>
      </div>
      <ul
        class="UnorderedListweb__UnorderedList-sc-1tb6ajo-0 List__StyledUnorderedList-bqf6r7-1 dbSpGu"
        data-blade-component="list"
        role="list"
      >
        <li
          class="ListItemElementweb__ListItemElement-sc-2pr284-0 ListItem__StyledListItem-sc-6bxu4b-0  hWTYHD"
          data-blade-component="list-item"
        >
          <div
            class="Boxweb__Box-xvyj85-0 jlsgap"
            display="flex"
          >
            <div
              class="Boxweb__Box-xvyj85-0 cXVkPc"
              display="flex"
            >
              <svg
                aria-hidden="true"
                data-blade-component="icon"
                fill="none"
                height="6px"
                viewBox="0 0 6 6"
                width="6px"
              >
                <circle
                  cx="3px"
                  cy="3px"
                  r="2.5px"
                  stroke="hsla(214, 18%, 69%, 1)"
                />
              </svg>
            </div>
            <p
              class="BaseTextweb__StyledBaseText-sc-1ob51ew-0 dxOnav Text__StyledText-wkh56f-0"
              color="surface.text.normal.lowContrast"
              data-blade-component="text"
              font-family="text"
              font-size="100"
              font-style="normal"
              font-weight="regular"
            >
              Level 2
            </p>
          </div>
          <ul
            class="UnorderedListweb__UnorderedList-sc-1tb6ajo-0 List__StyledUnorderedList-bqf6r7-1 dbSpGu"
            data-blade-component="list"
            role="list"
          >
            <li
              class="ListItemElementweb__ListItemElement-sc-2pr284-0 ListItem__StyledListItem-sc-6bxu4b-0  hWTYHD"
              data-blade-component="list-item"
            >
              <div
                class="Boxweb__Box-xvyj85-0 jlsgap"
                display="flex"
              >
                <div
                  class="Boxweb__Box-xvyj85-0 cXVkPc"
                  display="flex"
                >
                  <svg
                    aria-hidden="true"
                    data-blade-component="icon"
                    fill="none"
                    height="6px"
                    viewBox="0 0 6 6"
                    width="6px"
                  >
                    <rect
                      fill="hsla(214, 18%, 69%, 1)"
                      height="6px"
                      rx="1px"
                      width="6px"
                    />
                  </svg>
                </div>
                <p
                  class="BaseTextweb__StyledBaseText-sc-1ob51ew-0 dxOnav Text__StyledText-wkh56f-0"
                  color="surface.text.normal.lowContrast"
                  data-blade-component="text"
                  font-family="text"
                  font-size="100"
                  font-style="normal"
                  font-weight="regular"
                >
                  Level 3
                </p>
              </div>
            </li>
          </ul>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
